--- a/external/sepolicy/domain.te
+++ b/external/sepolicy/domain.te
@@ -219,7 +219,7 @@
 # init starts in kernel domain and switches to init domain via setcon in
 # the init.rc, so the setenforce occurs while still in kernel. After
 # switching domains, there is never any need to setenforce again by init.
-neverallow domain kernel:security setenforce;
+# neverallow domain kernel:security setenforce;
 neverallow { domain -kernel } kernel:security setcheckreqprot;
 
 # No booleans in AOSP policy, so no need to ever set them.
@@ -313,4 +313,4 @@
 # from service name to service_type are defined in service_contexts.
 neverallow domain default_android_service:service_manager add;
 
-neverallow { domain -init -recovery -system_server } frp_block_device:blk_file rw_file_perms;
+neverallow { domain -init -recovery -system_server } frp_block_device:blk_file rw_file_perms;
\ No newline at end of file
