From e0aaf3f89cee708186eccff026a5e9909e010e61 Mon Sep 17 00:00:00 2001
From: TheComputerGuy96 <vwestlife20@gmail.com>
Date: Mon, 23 May 2016 22:06:37 +0300
Subject: [PATCH] Actually make .patch file

---
 domain.te | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/domain.te b/domain.te
index 0f6c6da..b7aec3a 100644
--- a/domain.te
+++ b/domain.te
@@ -235,7 +235,7 @@ neverallow { domain -system_server } security_file:lnk_file { create setattr unl
 # init starts in kernel domain and switches to init domain via setcon in
 # the init.rc, so the setenforce occurs while still in kernel. After
 # switching domains, there is never any need to setenforce again by init.
-neverallow domain kernel:security setenforce;
+#neverallow domain kernel:security setenforce;
 neverallow { domain -kernel } kernel:security setcheckreqprot;
 
 # No booleans in AOSP policy, so no need to ever set them.
@@ -319,7 +319,7 @@ neverallow domain rootfs:file { create write setattr relabelto append unlink lin
 
 # Restrict context mounts to specific types marked with
 # the contextmount_type attribute.
-neverallow domain {fs_type -contextmount_type}:filesystem relabelto;
+neverallow domain {fs_type -contextmount_type -sdcard_posix}:filesystem relabelto;
 
 # Ensure that context mount types are not writable, to ensure that
 # the write to /system restriction above is not bypassed via context=
